vyper/ast/README.md

```diff
@@ -0,0 +1,127 @@
+This is used to handle discrepencies in how nodes are structured between different
+Python versions.
+* Fields listed in `_only_empty_fields`, if present within the Python AST, must
+be `None` or a `SyntaxException` is raised.
+* All other fields are ignored.
+
+Most Vyper nodes have an identical (or simplified) structure to their Python
+counterpart. Divergences are always handled in a Vyper node's `__init__` method,
+and explained in the docstring. To learn more about the structure of the nodes:
+
+* Green Tree Snakes: [Meet the Nodes](https://greentreesnakes.readthedocs.io/en/latest/nodes.html)
+
+### Traversing the AST
+
+Each node contains several methods that are used to traverse the AST. These methods
+can optionally include one or more filters in order to quickly search for children
+or parents that match a desired pattern.
+
+* `get_children`: Returns direct childen of the given node.
+* `get_descendants`: Returns direct and indirect childen of the given node.
+* `get_parent`: Returns a parent of the given node.
+* `get`: Recursive getter function for node attributes.
+
+To learn more about these methods, read their docstrings in the `VyperNode` class
+in [`nodes.py](nodes.py).
+
+## Design
+
+### `__slots__`
+
+Node classes in make use of `__slots__` for additional type safety. Denying the
+creation of `__dict__` in node classes ensures that unwanted fields cannot be
+silently included in the nodes.
+
+To learn more about `__slots__`:
+
+* Python Documentation: [`__slots__`](https://docs.python.org/3.8/reference/datamodel.html#slots)
+* Stack Overflow: [Usage of `__slots__`?](https://stackoverflow.com/a/28059785/11451521)
+
+### Interface Files (`.pyi`)
+
+This module makes use of Python interface files ("stubs") to aid in MyPy type
+annotation.
+
+Stubs share the same name as their source counterparts, with a `.pyi` extension.
+Whenever included, a stub takes precedence over a source file. For example, given
+the following file structure:
+
+```bash
+ast/
+  node.py
+  node.pyi
+```
+
+The type information in `node.pyi` is applied to `node.py`. Any types given in
+`node.py` are only there to aid code readability - they are ignored by the type
+checker.
+
+The following resources are useful for familiarizing yourself with stubs:
+
+* [MyPy: Stub Files](https://mypy.readthedocs.io/en/stable/stubs.html)
+* [PEP 484: Stub Files](https://www.python.org/dev/peps/pep-0484/#stub-files)
+
+## Integration
+
+All node classes are imported from `ast/nodes.py` into `ast/__init__.py`. If you are
+importing an AST node from another module, you should do so from the root `ast`
+package.
+
+For readability, `ast` should always be aliased as `vy_ast`.
+
+```python
+# Correct
+from vyper import ast as vy_ast
+from vyper.ast import VyperNode
+
+# Incorrect - could be mistaken for the builtin ast library
+from vyper import ast
+
+# Incorrect - only the functionality within vyper.ast is considered "public"
```

Comment:
We talked about whether underscoring made sense, and agreed that we wouldn't do it as long as this sort of distinction was understood, correct?

Reply:
Maybe add a note that it's not enforced, but highly suggested you not do this.

Reply:
Ooooh good call.  I'll add an explanation of when to prepend with underscores in #1871.

Reply:
https://github.com/vyperlang/vyper/pull/1871/files#diff-260f2647b0ee0405377642ee077b51f1R40